Add support for list, tuple and dict in numeric debugger

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143882

Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D67660049](https://our.internmc.facebook.com/intern/diff/D67660049)

jerryzh168 committed:
Add support for list, tuple and dict in numeric debugger

Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
     bfs_trace_with_node_process(ep, _assign_debug_handle)
 
 
+def _detach(x):
```
Add type annotations. `object` is fine if you do isinstance inside it and have a backup if it's not a tensor or a collection of tensors at all.

TypeVar might be helpful for this function

jerryzh168 replied:
I'm going to use Any

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
+def _tensor_shape_equals(x, y):
+    if isinstance(x, torch.Tensor) and isinstance(y, torch.Tensor):
+        print("x:", x.shape, " y:", y.shape)
+        print("equals:", x.shape == y.shape)
```
No prints committed

jerryzh168 committed:
Update on "Add support for list, tuple and dict in numeric debugger"


Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
+        print("x:", x.shape, " y:", y.shape)
+        print("equals:", x.shape == y.shape)
+        return x.shape == y.shape
+    elif isinstance(x, (list, tuple)):
```
check that y is a list or tuple too

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
+        detached = x
+    return detached
+
+def _tensor_shape_equals(x, y):
```
For these two functions, see if you can reuse `map_aggregate` from fx.node. It already handles all the tuple/dict branching stuff.
You can give it a callable for Tensor to do the detach or replace a Tensor with its shape for comparison

jerryzh168 replied:
I can use map_aggregate for `_detach`, but others does not help that much I think

jerryzh168 replied:
actually I'll revert this since it introduced a lot of typing errors..

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -71,6 +71,45 @@ def _assign_debug_handle(node: torch.fx.Node) -> None:
+        print(f"Comparing non Tensors: {x} and {y}, they must be equal")
+        return type(x) == type(y) and x == y
+
+def _loss_fn(loss, x, y):
```
Same here, reuse `map_aggregate` for this.
Also, type annotate `loss`

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -90,11 +129,10 @@ def __init__(
         self.nn_module_stack = nn_module_stack
         self.debug_handle = debug_handle
-        self.stats: List[torch.Tensor] = []
+        self.stats: List[Any] = []
```
Use `object` instead of `Any`

dulinriley commented on torch/ao/quantization/pt2e/_numeric_debugger.py:
```diff
@@ -163,26 +201,16 @@ class QuantizationComparisonResult:
-        return loss_function(
-            self.actual.to(dtype=torch.float32), self.ref.to(dtype=torch.float32)
-        )
+        return _loss_fn(loss_function, self.actual, self.ref)
```
You might want to leave in the `to(torch.float32)`, as most loss functions require floats.

Otherwise `self.loss(F.mse_loss)` fails for integer tensors

dulinriley commented on torch/testing/_internal/common_quantization.py:
```diff
@@ -2930,6 +2930,21 @@ def forward(self, x1, x2, x3, x4):
+
+        def forward(self, x):
+            x = self.conv1(x)
+            x1, x2 = x.split(2, dim=1)
```
Can you add a comment explaining why `split` is important? Is it because it returns a list/tuple of tensors?

jerryzh168 replied:
yes

dulinriley reviewed (approved):
Thanks for adding this functionality

jerryzh168 committed:
Update on "Add support for list, tuple and dict in numeric debugger"


Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).

jerryzh168 committed:
Update on "Add support for list, tuple and dict in numeric debugger"


Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D67660049](https://our.internmc.facebook.com/intern/diff/D67660049)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).

jerryzh168 committed:
Update on "Add support for list, tuple and dict in numeric debugger"


Summary:
Previously numeric debugger only supports torch.Tensor, this PR adds support for list, tuple and dict as well

Test Plan:
python test/test_quantization.py -k test_extract_results_from_loggers_list_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D67660049](https://our.internmc.facebook.com/intern/diff/D67660049)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67660049).

jerryzh168 commented:
@pytorchbot merge
